### PR TITLE
Store operation identifier in `IdLockMap` on exclusive lock

### PR DESCRIPTION
**Pull request by jbajic**
## Problem

Addresses https://github.com/neondatabase/neon/issues/7108

## Summary of changes

- As suggested in the issues a wrapper was added around `OwnedRwLockWriteGuard` called `IdentifierLock` that removes the operation currently holding the exclusive lock when it's dropped.
- The value in `IdLockMap` was extended to hold a pair of locks and operations that can be accessed and locked independently. 
- When requesting an exclusive lock besides returning the lock on that resource, an operation is changed if the lock is acquired.

## Checklist before requesting a review

- [x] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

**Review by VladLazar — COMMENTED** [review]
Sorry for the late review.

I find it a bit strange that we keep the `operation` in a separate rwlock. I'm guessing you did this because it works nicely for shared locks as well.

I would store the operation in the original rwlocks held by the `entities` map. When the lock is shared, there may be multiple operations, but I'd be ok with just tracking the latest one. Alternatively, we could use enum tags to identify operations and keep `smallvec::SmallVec` inside the lock to track all current lock holders.

**Comment by jbajic** [comment]
> Sorry for the late review.
> 
> I find it a bit strange that we keep the `operation` in a separate rwlock. I'm guessing you did this because it works nicely for shared locks as well.
> 
> I would store the operation in the original rwlocks held by the `entities` map. When the lock is shared, there may be multiple operations, but I'd be ok with just tracking the latest one. Alternatively, we could use enum tags to identify operations and keep `smallvec::SmallVec` inside the lock to track all current lock holders.

Thank you for taking a look at it.

My idea was that even when the exclusive lock was being held, the requesting operation could identify who was holding the ex lock at that moment, and so a lock for resource and another for operation. But since that is not needed I will remove it. My question is how will the identifier that is holding the lock be used?

**Comment by VladLazar** [comment]
> My question is how will the identifier that is holding the lock be used?

The idea is to get better observability around locking. Ideally, after a certain time trying to acquire the lock, we'd log what's currently holding up the lock, but that's not really possible since the lock is being held and we can't peek into it.

I think what would work reasonably well would be to:
* Issue a warning log when an acquire is taking too long (a few seconds?) - this won't include the operation id which is holding the lock, but will include the operation id of the thing that's trying to grab the lock
* When a lock is dropped, issue a warning if it was held for too long (a few seconds?) - this will include the operation which held the lock.

This gives us enough information to correlate long requests with the logs, although it doesn't help if we ever deadlock.

Let me know how much of this (if any) you plan to implement in this pr.

**Comment by VladLazar** [comment]
> Aha that makes sense, thanks for the explanation. I would be willing to implement both.

Very much appreciated! Please re-request review when you get around to it.

File: storage_controller/src/id_lock_map.rs
Comment (jbajic) [inline storage_controller/src/id_lock_map.rs:98]: I am not sure what would be the best way to test these two trace functions... 

I have looked into [tracing_test](https://docs.rs/tracing-test/latest/tracing_test/) but it seems just a bit too heavy to use in unit tests and slows it down for the `LOCK_TIMEOUT_ALERT_THRESHOLD` duration. 
On the other hand, I see that there are logging tests in `test_runner/regress/test_logging.py` but here I am not sure how would I cause these locks to hold for the duration to escalate to the point of being logged consistently.

File: storage_controller/src/id_lock_map.rs
Comment (VladLazar) [inline storage_controller/src/id_lock_map.rs:27]: nit: could do `self.start.elapsed()` instead

File: storage_controller/src/id_lock_map.rs
Comment (VladLazar) [inline storage_controller/src/id_lock_map.rs:109]: nit: `start.elapsed()`

File: storage_controller/src/id_lock_map.rs
Comment (VladLazar) [inline storage_controller/src/id_lock_map.rs:132]: Let's include the `key` in this log line. It's fine to clone it.

File: storage_controller/src/id_lock_map.rs
Comment (VladLazar) [inline storage_controller/src/id_lock_map.rs:112]: Let's include the key in this log line. It's fine to clone it.

File: storage_controller/src/id_lock_map.rs
Comment (VladLazar) [reply storage_controller/src/id_lock_map.rs:98]: How about writing a python regress test. You could:
1. Add a `pausable_failpoint` to one of the exclusive operations done by the storage controller
2. Configure the failpoint
3. Send an HTTP request which triggers the failpoint
4. Send another HTTP request which triggers an attempt to grab the lock
5. At the end of the test validate that the warning were printed

`NeonStorageController` python class will be useful. Also, the tests in `test_storage_controller.py` will be useful as examples. The `NeonPageserver` python class already has a `log_contains` and you'll need to add a similar one to `NeonStorageController`.

I'm concious this is rather involved, but such a test would be very valuable. Let me know if you need help (e.g. running the tests locally).

**Review by VladLazar — COMMENTED** [review]
This looks very nice. Good job!

File: test_runner/fixtures/neon_fixtures.py
Comment (VladLazar) [inline test_runner/fixtures/neon_fixtures.py:2290]: Let's try to avoid duplicating this code. Could you please abstract this stuff into a separate mixin class and make both `NeonStorageController` and `NeonPageserver` inherit from it? Alternatively, you could lift what's needed in free functions and have a lightweight wrapper in each class.

**Review by VladLazar — COMMENTED** [review]
Nice test. Triggering ci

File: test_runner/fixtures/neon_fixtures.py
Comment (VladLazar) [inline test_runner/fixtures/neon_fixtures.py:1953]: nit: maybe `LogUtils` would be a better name?

File: test_runner/fixtures/neon_fixtures.py
Comment (VladLazar) [inline test_runner/fixtures/neon_fixtures.py:1955]: nit: this comment doesn't make a lot of sense to me. How about: `A mixin class which provides utilities for inspecting the logs of a service`

**Comment by jbajic** [comment]
> ### 2838 tests run: 2677 passed, 40 failed, 121 skipped ([full report](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html))
> #### Failures on Postgres 16
> * `test_sharding_split_failures[failure2]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/d0760c7a8849c578), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/7e2121851b8ba270)
> * `test_sharding_split_failures[failure0]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/19346e4fa6f8466e), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/842396f2a4499d63)
> * `test_sharding_split_smoke`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/d14c39141e376943), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/ee4618f152f7fc6b)
> * `test_sharding_split_failures[failure3]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/6fad1865370daeca)
> * `test_sharding_split_failures[failure1]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/a342ae71367ba4ad)
> * `test_sharding_split_failures[failure19]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/a3efbcb2c663904f)
> * `test_sharding_split_failures[failure20]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/1ed5c4e415899620)
> * `test_tenant_delete_concurrent`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/3556ed71f2d69272a7014df6dcb02317/aa87fb88c466a63), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/3556ed71f2d69272a7014df6dcb02317/d67cfbfae7f4b1e5)
> 
> #### Failures on Postgres 15
> * `test_sharding_split_failures[failure0]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/27e644071ac2ada7), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/a9c89b4d00e7af42)
> * `test_sharding_split_failures[failure2]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/52fd5ac8e503e6e), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/38e74b1184851f99)
> * `test_sharding_split_failures[failure19]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/bd515c55cdae29d), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/e828cf687a1d54ce)
> * `test_sharding_split_smoke`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/2e5b9c43ebb4449c), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/a3f34d0cca9462bf)
> * `test_sharding_split_failures[failure3]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/45ea8cc3b5975cc2)
> * `test_sharding_split_failures[failure1]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/4d94c8e507bfaef9)
> * `test_sharding_split_failures[failure20]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/5c0721e2287cc191)
> * `test_tenant_delete_concurrent`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/3556ed71f2d69272a7014df6dcb02317/2704ad6256e3608d), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/3556ed71f2d69272a7014df6dcb02317/bd31f53a16f47cb5)
> 
> #### Failures on Postgres 14
> * `test_sharding_split_failures[failure0]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/39d17636d3ba086a), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/64067d07b201c8d3)
> * `test_sharding_split_failures[failure2]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/6d959b3f67e8dfe6), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/b9607de072c124a5)
> * `test_sharding_split_failures[failure1]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/2b24c45d0ec52ffe), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/a93aca6258e600b7)
> * `test_sharding_split_failures[failure19]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/83f0f92fa0cef731), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/7aff736da2c2c51)
> * `test_sharding_split_smoke`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/7a4876400c329cdc), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/59b8eda73fcb1237)
> * `test_sharding_split_failures[failure3]`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/c9a55e75a4fa46cd), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/5ea6f67b6fc882ea)
> * `test_sharding_split_failures[failure20]`: [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/140824de6e814b5b1ae2b622c3f67840/161badc923e67c1f)
> * `test_tenant_delete_concurrent`: [release](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/3556ed71f2d69272a7014df6dcb02317/75faef276db1185f), [debug](https://neon-github-public-dev.s3.amazonaws.com/reports/pr-7542/8880627534/index.html#suites/3556ed71f2d69272a7014df6dcb02317/ee76084c93e21536)
> 
> ```
> # Run all failed tests locally:
> scripts/pytest -vv -n $(nproc) -k "test_sharding_split_failures[release-pg14-failure0] or test_sharding_split_failures[debug-pg14-failure0] or test_sharding_split_failures[release-pg14-failure2] or test_sharding_split_failures[debug-pg14-failure2] or test_sharding_split_failures[release-pg14-failure1] or test_sharding_split_failures[debug-pg14-failure1] or test_sharding_split_failures[release-pg14-failure19] or test_sharding_split_failures[debug-pg14-failure19] or test_sharding_split_smoke[release-pg14] or test_sharding_split_smoke[debug-pg14] or test_sharding_split_failures[release-pg14-failure3] or test_sharding_split_failures[debug-pg14-failure3] or test_sharding_split_failures[debug-pg14-failure20] or test_tenant_delete_concurrent[release-pg14] or test_tenant_delete_concurrent[debug-pg14] or test_sharding_split_failures[release-pg15-failure0] or test_sharding_split_failures[debug-pg15-failure0] or test_sharding_split_failures[release-pg15-failure2] or test_sharding_split_failures[debug-pg15-failure2] or test_sharding_split_failures[release-pg15-failure19] or test_sharding_split_failures[debug-pg15-failure19] or test_sharding_split_smoke[release-pg15] or test_sharding_split_smoke[debug-pg15] or test_sharding_split_failures[debug-pg15-failure3] or test_sharding_split_failures[debug-pg15-failure1] or test_sharding_split_failures[debug-pg15-failure20] or test_tenant_delete_concurrent[release-pg15] or test_tenant_delete_concurrent[debug-pg15] or test_sharding_split_failures[release-pg16-failure2] or test_sharding_split_failures[debug-pg16-failure2] or test_sharding_split_failures[release-pg16-failure0] or test_sharding_split_failures[debug-pg16-failure0] or test_sharding_split_smoke[release-pg16] or test_sharding_split_smoke[debug-pg16] or test_sharding_split_failures[debug-pg16-failure3] or test_sharding_split_failures[debug-pg16-failure1] or test_sharding_split_failures[debug-pg16-failure19] or test_sharding_split_failures[debug-pg16-failure20] or test_tenant_delete_concurrent[release-pg16] or test_tenant_delete_concurrent[debug-pg16]"
> ```
> 
> Flaky tests (2)
> #### Test coverage report is not available
> The comment gets automatically updated with the latest test results
> [07643a7](https://github.com/neondatabase/neon/commit/07643a75c0c3486a49d12f9d6e9ca3a685d1ac5b) at 2024-04-29T15:24:34.111Z :recycle:

I see that most of these tests have warnings on locks being held for long period of time, will add these errors as allowed then

**Comment by VladLazar** [comment]
> I see that most of these tests have warnings on locks being held for long period of time, will add these errors as allowed then

Means it's working which is great! Didn't expect for these locks to be held for this long, interesting.

File: test_runner/fixtures/neon_fixtures.py
Comment (VladLazar) [inline test_runner/fixtures/neon_fixtures.py]: Interesting. So the issue was that we returned a `LogCursor` which wasn't pointing to the right line when we eventually found the match?

File: test_runner/regress/test_sharding.py
Comment (VladLazar) [inline test_runner/regress/test_sharding.py:226]: Instead of allow listing these logs, let's bump `LOCK_TIMEOUT_ALERT_THRESHOLD` to be equal to `RECONCILER_TIMEOUT`. Most (all?) operations that grab these locks speak to the pageserver and expect it to perform some IO, so it makes sense for it to be longer at this point.

Comment (VladLazar) [reply test_runner/regress/test_sharding.py:226]: Same for `test_sharding_split_failures` and you'll have to tweak the failpoint in `test_lock_time_tracing`.

File: test_runner/fixtures/neon_fixtures.py
Comment (jbajic) [reply test_runner/fixtures/neon_fixtures.py]: Ah yes, but that was my fault since when I copied the code I messed it up. This is the same code as on master

File: test_runner/regress/test_sharding.py
Comment (jbajic) [reply test_runner/regress/test_sharding.py:226]: Makes sense, will do that

Comment (jbajic) [reply test_runner/regress/test_sharding.py:226]: Did you mean `SHORT_RECONCILE_TIMEOUT`(5s) or `RECONCILE_TIMEOUT`(30s)?

Comment (VladLazar) [reply test_runner/regress/test_sharding.py:226]: `RECONCILE_TIMEOUT` (30s)

File: test_runner/fixtures/neon_fixtures.py
Comment (VladLazar) [reply test_runner/fixtures/neon_fixtures.py]: Ah I see. Missed that in the review.

File: storage_controller/src/id_lock_map.rs
Comment (VladLazar) [inline storage_controller/src/id_lock_map.rs:58]: Would be nice to do something similar via a `WrappedReadGuard` type for shared locks, but it can be a follow-up.

**Review by VladLazar — COMMENTED** [review]
Looking good. Let's run CI one more time.

File: storage_controller/src/id_lock_map.rs
Comment (jbajic) [reply storage_controller/src/id_lock_map.rs:58]: I would like to take over the follow-up too

Comment (VladLazar) [reply storage_controller/src/id_lock_map.rs:58]: Sounds good. Feel free to open a pr. I'll merge this when CI stops flaking.

**Review by VladLazar — APPROVED** [review]

File: storage_controller/src/id_lock_map.rs
Comment (VladLazar) [reply storage_controller/src/id_lock_map.rs:58]: And thank you for the contribution!

**Pull request merged**
